### ISSUE TITLE
SCM Update and Job Templates marked as failed when one of inventory sync failed during another inventory sync is running

### ISSUE THREAD

**Issue by hiyokotaisa** [ansible/awx#4418]
##### ISSUE TYPE
 - Bug Report

##### SUMMARY

SCM update has been marked as failed with the error `inventory_update failed for <INVENTORY NAME> - <INVENTORY SOURCE NAME> with ID <INVENTORY SYNC JOBID>` even if there is no error in the SCM update job.

This will happen when the `Inventory Update X` is launched and running and `Inventory Update Y` is failed while `Inventory Update X` is still running. This situation will occur when there is slow inventory source (vCenter with massive host entries for example).

Also, even if the project update is succeeded when I launch the Job Template, the job will be marked as failed after a project update for another Job Template has been failed. It should be marked as success since the project update for Job Template is succeeded.

Here is the video for reproducing the issue:
![tower](https://user-images.githubusercontent.com/4246532/62445287-d9e4ae00-b79a-11e9-85e5-19052564178e.gif)

##### STEPS TO REPRODUCE

1. Create 2 Job Template as follows:
```
Demo Project
        ┣ Test Job Template X
        ┗ Test Job Template Y
```

Test Job Template X:
![Screenshot from 2019-08-05 13-49-04](https://user-images.githubusercontent.com/4246532/62439763-0cd17680-b788-11e9-9830-ad2aab8174c1.png)

Test Job Template Y:
![Screenshot from 2019-08-05 13-49-12](https://user-images.githubusercontent.com/4246532/62439775-18bd3880-b788-11e9-9bfd-41226a466ffd.png)

2. Create Inventories for Job Template X and Y with Custom Inventory Scripts to reproduce the issue.

For Test Job Template X:
```
Test Inventory X
        ┗ Test Custom Inventory Source X
                ┗ Test Custom Inventory X
```
Here is the Test Custom Inventory X to sleep 60sec then add hosts and groups:
```
#!/usr/bin/env python

import json
import sys
import time


SLEEP = 60

inv = {}
argv = sys.argv
opts = [o for o in argv if o.startswith('--')]

if '--host' in opts:
    pass
else:
    inv = dict(_meta=dict(hostvars=dict()), all=dict(children=list()))
    # dummy data
    inv["groupA"] = { "hosts": ["test01"], "vars": {} }
    inv["groupB"] = { "hosts": ["test11"], "vars": {} }
    inv["_meta"]["hostvars"] = { "test01": { "ansible_connection": "local" }, "test11": { "ansible_connection": "local" }}

time.sleep(SLEEP)
print(json.dumps(inv))

#
# [EOF]
#
```

For Test Job Template Y:
```
Test Inventory Y
        ┗ Test Custom Inventory Source Y
                ┗ Test Custom Inventory Y
```

Here is the Test Custom Inventory Y to sleep 30sec then failed on purpose:
```
#!/usr/bin/env python

import json
import sys
import time


SLEEP = 30

inv = {}
argv = sys.argv
opts = [o for o in argv if o.startswith('--')]

if '--host' in opts:
    pass
else:
    inv = dict(_meta=dict(hostvars=dict()), all=dict(children=list()))
    # dummy data
    inv["groupA"] = { "hosts": ["test00"], "vars": {} }
    inv["groupB"] = { "hosts": ["test10"], "vars": {} }
    inv["_meta"]["hostvars"] = { "test00": { "ansible_connection": "local" }, "test10": { "ansible_connection": "local" }}

time.sleep(SLEEP)
raise ValueError("Testing for CustomInventoryScriptN")
print(json.dumps(inv))

#
# [EOF]
#
```

3. Launch `Job Template X` first then launch `Job Template Y` about 30 sec later.

##### EXPECTED RESULTS

For Job Template X, SCM Update, Inventory Update and Job Template are marked as success.
For Job Template Y, Inventory Update fails due to error but SCM update succeeded.

##### ACTUAL RESULTS

For Job Template X, SCM Update and Inventory Update are marked as success but Job Template has been failed with `The project revision for this job template is unknown due to a failed update.` 

When I launch the `Job Template X`, I can see the project update is success. But after the `Inventory Update Y` has been failed, `Job Template X` and all of following Job Templates are failed with the following error:
![Screenshot from 2019-08-05 14-45-56](https://user-images.githubusercontent.com/4246532/62441614-d4359b00-b78f-11e9-8149-c9c93f72ca07.png)

Please refer the video I posted on the SUMMARY section too. 

For Job Template Y, Both of Inventory Update and SCM update fail.

SCM Update has been finished without error but marked as failed with the following message:
```
inventory_update failed for Test Inventory Y - Test Custom Inventory Source Y with ID 35
```
![Screenshot from 2019-08-05 14-21-37](https://user-images.githubusercontent.com/4246532/62440767-7ce1fb80-b78c-11e9-982d-b2a154b5a058.png)

I cannot see any issues in the process of SCM update.

**Comment by AlanCoding** [comment]
What update-on-launch options are selected for the projects & inventory sources mentioned in these steps?

**Comment by hiyokotaisa** [comment]
@AlanCoding Thanks for the comment, Update on Launch is enabled on Project but it is disabled on Inventory Source

**Comment by ryanpetrello** [comment]
@AlanCoding @fosterseth I wonder if this is still applicable given our recent inventory task manager changes.

**Comment by fosterseth** [comment]
@ryanpetrello @AlanCoding This is still the behavior, even after the recent task manager changes.

If an inventory source update fails it will auto-fail the project update too, even if that project update already ran successfully. As a result -- other jobs that depended on that project update will also fail even if those jobs had nothing to do with the failed inventory source update.

I can see where this is a problem. If someone spawn 30 jobs with Inventory A and 30 jobs with Inventory B, all based on the same project update, and then inventory A update fails, it will bring down all 60 jobs not just the 30.

we can alter this line https://github.com/ansible/awx/blob/devel/awx/main/tasks.py#L587

to be `if instance.celery_task_id != task_id and not instance.cancel_flag and not instance.status == "successful":`

basically, don't fail a task if it already ran successfully.

**Comment by ryanpetrello** [comment]
@fosterseth sounds like an opportunity for a PR 😄

**Comment by kdelee** [comment]
I used above examples to create two jobs that relied on the same project update, set to update on launch, but have two distinct inventories, one that always fails on update and another that always succeeds.

I worked through the reproducer as well as various other combinations of running the jobs and inventory and project updates.  In no case did I see a failing inventory update for JT Y that uses inventory Y that fails effect the shared project updates that JTY and JT X use, nor did it effect jobs from JT X that uses inventory X that succeed.

Calling this good and going to close. Thanks for this amazingly detailed report @hiyokotaisa !!!!
